src/lib/comms/sol-socket-impl-zephyr.c

```diff
@@ -0,0 +1,383 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+
+#include <atomic.h>
+#include <net/ip_buf.h>
+#include <sections.h>
+
+#include "sol-buffer.h"
+#include "sol-log.h"
+#include "sol-mainloop.h"
+#include "sol-mainloop-zephyr.h"
+#include "sol-socket-impl.h"
+#include "sol-vector.h"
+#include "sol-network-util.h"
+
+struct sol_socket_zephyr {
+    struct sol_socket base;
+
+    struct {
+        bool (*cb)(void *data, struct sol_socket *s);
+        const void *data;
+    } read, write;
+
+    struct sol_timeout *write_timeout;
+
+    struct net_context *unicast_context;
+    struct sol_ptr_vector mcast_contexts;
+    struct sol_ptr_vector bufs;
+
+    struct nano_sem lock;
+    bool read_available;
+};
+
+static struct sol_ptr_vector bound_sockets = SOL_PTR_VECTOR_INIT;
+
+static void
+socket_read_available(void *data)
+{
+    struct sol_socket_zephyr *s = data;
+    uint16_t count;
+
+    nano_task_sem_take(&s->lock, TICKS_UNLIMITED);
+    count = sol_ptr_vector_get_len(&s->bufs);
+    nano_task_sem_give(&s->lock);
+
+    if (s->read.cb)
+        while (count--)
+            s->read.cb((void *)s->read.data, &s->base);
+}
+
+static void
+socket_signal_mainloop(struct sol_socket_zephyr *s, struct net_buf *buf)
+{
+    struct mainloop_event me = {
+        .cb = socket_read_available,
+        .data = s
+    };
+
+    nano_fiber_sem_take(&s->lock, TICKS_UNLIMITED);
+
+    sol_ptr_vector_append(&s->bufs, buf);
+
+    if (!s->read_available) {
+        sol_mainloop_event_post(&me);
+        s->read_available = true;
```

Comment:
On Mon, 28 Mar 2016 13:59:05 -0700, Gustavo Lima Chaves wrote:
> > +
> > +static void
> > +socket_signal_mainloop(struct sol_socket_zephyr *s, struct net_buf *buf)
> > +{
> > +    struct mainloop_event me = {
> > +        .cb = socket_read_available,
> > +        .data = s
> > +    };
> > +
> > +    nano_fiber_sem_take(&s->lock, TICKS_UNLIMITED);
> > +
> > +    sol_ptr_vector_append(&s->bufs, buf);
> > +
> > +    if (!s->read_available) {
> > +        sol_mainloop_event_post(&me);
> > +        s->read_available = true;
> 
> This is never checked elsewhere.

The problem is that I'm not setting it back to false when read.

> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/solettaproject/soletta/pull/1743/files/efa9fcb413c3587b7ec630f3daa689a4a8161937..9d8f3e6d9e0fff96da2ac5d0471b21e0794fcef4#r57632731